### PR TITLE
Add S3 multipart copy and use it in Nexus ship

### PR DESCRIPTION
**Pull request by olivergrabinski**

### CODE REVIEW

File: delta/plugins/storage/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/storage/storages/operations/s3/client/S3StorageClientImpl.scala [hunk 1]
```diff
@@ -76,6 +70,56 @@ final private[client] class S3StorageClientImpl(client: S3AsyncClientOp[IO], val
         .build()
     )
 
+  def copyObjectMultiPart(
+      sourceBucket: String,
+      sourceKey: String,
+      destinationBucket: String,
+      destinationKey: String
+  ): IO[CompleteMultipartUploadResponse] = {
+    val partSize = 5L * 1024L * 1024L // 5MB
```
Comment (imsdu) [inline delta/plugins/storage/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/storage/storages/operations/s3/client/S3StorageClientImpl.scala:79]: That may be a little low no ?

Comment (imsdu) [reply delta/plugins/storage/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/storage/storages/operations/s3/client/S3StorageClientImpl.scala:79]: Also some files are empty, we have to make sure it works, maybe for files lower than the part size we should not use multipart ?

File: delta/plugins/storage/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/storage/storages/operations/s3/client/S3StorageClientImpl.scala [hunk 2]
```diff
@@ -76,6 +70,56 @@ final private[client] class S3StorageClientImpl(client: S3AsyncClientOp[IO], val
...
+        val uploadParts = (0L until objectSize by partSize).zipWithIndex.map { case (start, partNumber) =>
+          val lastByte              = Math.min(start + partSize - 1, objectSize - 1)
+          val uploadPartCopyRequest = UploadPartCopyRequest.builder
+            .sourceBucket(sourceBucket)
+            .sourceKey(sourceKey)
+            .destinationBucket(destinationBucket)
+            .destinationKey(destinationKey)
+            .uploadId(createMultipartUploadResponse.uploadId)
+            .partNumber(partNumber + 1)
+            .copySourceRange(s"bytes=$start-$lastByte")
```
Comment (dantb) [inline delta/plugins/storage/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/storage/storages/operations/s3/client/S3StorageClientImpl.scala:97]: This is fiddly, will S3 complain if the parts are not mutually consistent? I suppose the tests are good and pass it should be fine!

**Review by imsdu — APPROVED** [review]

File: delta/plugins/storage/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/storage/storages/operations/s3/client/S3StorageClientImpl.scala (hunk 2 above)
Comment (olivergrabinski) [reply delta/plugins/storage/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/storage/storages/operations/s3/client/S3StorageClientImpl.scala:97]: You're right, but as long as we don't use the high-level api (TransferManager) then I'm failing to see a non-fiddly way, but it's possible I'm wrong

**Pull request merged**